### PR TITLE
fix(arborist): allow registry bundle inspection in dry runs

### LINKED ISSUE #9800: [BUG] `--dry-run` fails with EALLOWREMOTE on registry tarballs for packages that declare bundleDependencies
### Is there an existing issue for this?

- [x] I have searched the existing issues

The closest reports have a different cause: #9796 (registry serving tarballs from a path outside the configured registry prefix) and the closed exemption fixes #9494 / #9548 / #9509. This one happens against the default `registry.npmjs.org` and only under `--dry-run`. #9187 is a separate `--dry-run` bug (side effects / output).

### This issue exists in the latest npm version

- [x] I am using the latest npm

### This is not just a request to bump a dependency for a CVE

- [x] This is not solely a request to bump a dependency for a CVE

### Current Behavior

With npm 12's default `allow-remote=none`, `npm install --dry-run` refuses a registry-hosted tarball when the target package — or a package in its graph — declares `bundleDependencies`:

```console
$ npm i --dry-run npm@11.10.0
npm error code EALLOWREMOTE
npm error Fetching packages of type "remote" have been disabled
npm error Refusing to fetch "https://registry.npmjs.org/npm/-/npm-11.10.0.tgz"
```

The identical command without `--dry-run` succeeds. The refused URL is a `registry.npmjs.org` tarball and `registry` is the default, so the registry-tarball exemption should apply.

| command | result |
| --- | --- |
| `npm i --dry-run npm@11.10.0` | `EALLOWREMOTE` |
| `npm i npm@11.10.0` | added 1 package |
| `npm i --dry-run @tailwindcss/oxide@4.3.3` | `EALLOWREMOTE` |
| `npm i @tailwindcss/oxide@4.3.3` | added 2 packages |
| `npm i --dry-run lodash@4.17.21` (control, no `bundleDependencies`) | added 1 package |

`@tailwindcss/oxide` is the realistic case: it bundles nothing itself, but its optional `@tailwindcss/oxide-wasm32-wasi` dep declares `bundleDependencies` — and since that package is `cpu: ["wasm32"]`, a real install skips it entirely, which is presumably why only the dry-run path trips.

Practical impact: `npm update --dry-run` to preview a bump fails in any project whose tree contains such a package (anything on Tailwind 4, for instance), while the real `npm update` completes fine. In a monorepo that makes the preview unusable without `--allow-remote=all`, which is exactly the guard one doesn't want to relax to inspect a change.

Not caused by `min-release-age`: reproduces with `--min-release-age=0` explicitly passed.

### Expected Behavior

`--dry-run` should apply the same registry-tarball `allow-remote` exemption a real install does. Declaring `bundleDependencies` shouldn't cause a registry spec to be classified as type `remote`.

### Steps To Reproduce

```console
$ mkdir /tmp/eallowremote && cd /tmp/eallowremote && npm init -y

# both fail with EALLOWREMOTE
$ npm i --dry-run npm@11.10.0
$ npm i --dry-run @tailwindcss/oxide@4.3.3

# both succeed
$ npm i npm@11.10.0
$ npm i @tailwindcss/oxide@4.3.3

# control: no bundleDependencies, dry-run is fine
$ npm i --dry-run lodash@4.17.21
```

### Environment

- npm: 12.0.1
- Node.js: 24.18.0
- OS Name: Linux 7.0.11 (x86_64)
- System Model Name: n/a
- npm config:

```ini
; user config (registry auth lines omitted)
min-release-age = 7
min-release-age-exclude = ["@fullbeaker/*"]
registry = "https://registry.npmjs.org/"

; relevant npm 12 defaults, not overridden
allow-remote = "none"
allow-git = "none"
```

### PR DESCRIPTION
**Pull request by yszgj-30**
## Summary

- share the existing registry-resolved tarball validation between reification and ideal-tree building
- apply the registry tarball exemption when a complete ideal tree inspects bundled dependencies
- add a regression test for `allowRemote=none`

## Root cause

Complete ideal-tree builds crack open tarballs for packages with bundled dependencies. That path passed the registry tarball URL directly to `pacote.extract()`, which classified it as a remote dependency and rejected it under the default `allowRemote=none` policy.

The fix only overrides that policy after the existing origin and registry-path checks confirm that the effective URL is registry-mediated.

## Validation

- reproduced `npm install --dry-run npm@11.10.0 --allow-remote=none` failing with `EALLOWREMOTE` before the fix
- verified the same command exits 0 after the fix
- passed the focused ideal-tree regression test
- passed the existing 18-test registry URL and `allowRemote` reify group
- passed `node . run lint -w @npmcli/arborist`

Fixes #9800

### CODE REVIEW

**Comment by MikeMcC399** [comment]
@martinrrm 

Could this be prioritized? Recent updates to the widely used [Renovate bot](https://github.com/renovatebot/renovate/releases) have caused the underlying issue to be triggered which causes Renovate failures - see https://github.com/npm/cli/issues/9800#issuecomment-5712060083

Renovate updates to an out-of-the-box installation created with `npx create-next-app@latest` for Next.js (see https://nextjs.org/docs/app/getting-started/installation) fail because of the related issue.

File: workspaces/arborist/lib/arborist/build-ideal-tree.js [hunk 1]
```diff
@@ -1028,6 +1029,8 @@ This is a one-time fix-up, please be patient...
           Arborist,
           resolved: node.resolved,
           integrity: node.integrity,
+          ...(isRegistryResolvedTarball(node, this.options) ?
```
Comment (martinrrm) [inline workspaces/arborist/lib/arborist/build-ideal-tree.js:1032]: This exemption can bypass `allowRemote: 'none'` when `replaceRegistryHost` is a full URL with a path. `isRegistryResolvedTarball()` treats the URL as rewriteable, but `pacote.extract()` receives the unchanged `node.resolved`; Pacote’s host comparison does not perform the same rewrite in this configuration. As a result, complete/dry-run tree construction may contact the original remote host. Could this pass the `registryResolved(...)` URL to `pacote.extract()`, as reification does, or base the exemption on the URL Pacote will actually fetch? A regression test with a path-bearing  `replaceRegistryHost` would cover this mismatch.

File: workspaces/arborist/lib/arborist/build-ideal-tree.js
Comment (martinrrm) [inline workspaces/arborist/lib/arborist/build-ideal-tree.js:1026]: Use the rewritten URL whenever granting the registry exemption, both as the extraction spec and the `resolved` option:

```js
const {
  isRegistryResolvedTarball,
  registryResolved,
} = require('../registry-resolved.js')
```

Then update the extraction block:

```js
await cacache.tmp.withTmp(this.cache, opt, async path => {
  const isRegistry = isRegistryResolvedTarball(node, this.options)
  const resolved = isRegistry
    ? registryResolved(node.resolved, this.options)
    : node.resolved

  await pacote.extract(resolved, path, {
    ...opt,
    Arborist,
    resolved,
    integrity: node.integrity,
    ...(isRegistry ? { allowRemote: 'all' } : {}),
  })

  await new Arborist({ ...this.options, path })
    .loadActual({ root: node, ignoreMissing: true })
})
```

This keeps non-exempt dependencies unchanged while making the registry exemption and extraction use the same rewritten URL.

Add a regression test with `complete: true`, `allowRemote: 'none'`, and a full-URL `replaceRegistryHost` containing a path. It should assert that extraction targets the configured mirror, not the original host.

| Case | Required assertion |
|---|---|
| Full-URL `replaceRegistryHost` containing a path | Extraction requests the rewritten mirror URL, never the original host. |
| Path-prefixed registry with default host replacement | Extraction preserves the configured registry path. |
| Non-registry remote dependency | Still rejects with `EALLOWREMOTE` without fetching. |
| Same-origin tarball outside the registry path, with replacement disabled | Still rejects with `EALLOWREMOTE` without fetching. |

**Review by martinrrm — CHANGES_REQUESTED** [review]

**Comment by martinrrm** [comment]
@yszgj-30 Thanks for opening the PR, I left some comments, but this looks great! Let us know if you can continue working on this fix.
